ooni/nettests/blocking/whatsapp.py

```diff
@@ -0,0 +1,418 @@
+    followRedirects = True
+    usageOptions = UsageOptions
+
+    def setUp(self):
+        # We need more time to ensure all the endpoints can timeout
+        self.timeout = 16 * 20
+
+    @defer.inlineCallbacks
+    def test_registration_server(self):
+        self.report['registration_server_failure'] = None
+        self.report['registration_server_status'] = None
+
+        url = 'https://v.whatsapp.net/v2/register'
+        # Ensure I get back:
+        # {"status": "fail", "reason": "missing_param", "param": "code"}
+
+        try:
+            yield self.doRequest(url, 'GET')
+        except Exception as exc:
+            failure_string = failureToString(defer.failure.Failure(exc))
+            log.err("Failed to contact the registration server %s" % failure_string)
+            self.report['registration_server_failure'] = failure_string
+            self.report['registration_server_status'] = 'blocked'
+            defer.returnValue(None)
+
+        log.msg("Successfully connected to registration server!")
+        self.report['registration_server_status'] = 'ok'
+
+    @defer.inlineCallbacks
+    def _test_whatsapp_web(self, url):
+        try:
+            response = yield self.doRequest(url, 'GET')
+        except Exception as exc:
+            failure_string = failureToString(defer.failure.Failure(exc))
+            log.err("Failed to connect to whatsapp web %s" % failure_string)
+            self.report['whatsapp_web_failure'] = failure_string
+            self.report['whatsapp_web_status'] = 'blocked'
+            defer.returnValue(None)
+
+        title = extractTitle(response.body).strip()
+        if title != "WhatsApp Web":
+            self.report['whatsapp_web_status'] = 'blocked'
+
+    @defer.inlineCallbacks
+    def test_whatsapp_web(self):
+        self.report['whatsapp_web_failure'] = None
+        self.report['whatsapp_web_status'] = None
+
+        yield self._test_whatsapp_web('https://web.whatsapp.com/')
+        yield self._test_whatsapp_web('http://web.whatsapp.com/')
+        if self.report['whatsapp_web_status'] != 'blocked':
+            self.report['whatsapp_web_status'] = 'ok'
+
+    def _test_connect_to_port(self, address, port):
+        result = {
+            'ip': address,
+            'port': port,
+            'status': {
+                'success': None,
+                'failure': None
+            }
+        }
+        point = TCP4ClientEndpoint(reactor, address, port, timeout=10)
+        d = point.connect(TCPConnectFactory())
+        @d.addCallback
+        def cb(p):
+            result['status']['success'] = True
+            result['status']['failure'] = False
+            self.report['tcp_connect'].append(result)
+
+        @d.addErrback
+        def eb(failure):
+            result['status']['success'] = False
+            result['status']['failure'] = failureToString(failure)
+            self.report['tcp_connect'].append(result)
+            return failure
+
+        return d
```

Comment:
So I thought a bit more about this and also about measuring latencies and I think this would require a bit of thought to come up with a good spec for the data format used to implement it.

I wrote down some notes on this here: https://github.com/TheTorProject/ooni-probe/issues/689.

I would say we defer this for the moment to some bigger PR in the future as it will also require breaking the data format of past tests and probably implementing some translation code to old to new format.

I would put it under this: https://github.com/TheTorProject/ooni-spec/issues/74